crypto/s2n_hkdf.c

```diff
@@ -175,7 +175,101 @@ static int s2n_libcrypto_hkdf(struct s2n_hmac_state *hmac, s2n_hmac_algorithm al
 
     return S2N_SUCCESS;
 }
+
+bool s2n_libcrypto_supports_hkdf()
+{
+    return true;
+}
+
+#elif S2N_OPENSSL_VERSION_AT_LEAST(3, 0, 0)
+
+    #include "crypto/s2n_kdf.h"
+
+static S2N_RESULT s2n_hkdf_kdf(struct s2n_hmac_state *hmac, s2n_hmac_algorithm alg,
+        const struct s2n_blob *salt, const struct s2n_blob *key, const struct s2n_blob *info,
+        struct s2n_blob *output, int mode)
+{
+    /* As an optimization, we should be able to fetch and cache this EVP_KDF*
+     * once when s2n_init is called.
+     */
+    DEFER_CLEANUP(EVP_KDF *hkdf_impl = EVP_KDF_fetch(NULL, "HKDF", NULL),
+            EVP_KDF_free_pointer);
+    RESULT_ENSURE(hkdf_impl, S2N_ERR_PRF_INVALID_ALGORITHM);
+
+    DEFER_CLEANUP(EVP_KDF_CTX *hkdf_ctx = EVP_KDF_CTX_new(hkdf_impl),
+            EVP_KDF_CTX_free_pointer);
+    RESULT_ENSURE_REF(hkdf_ctx);
+
+    const EVP_MD *digest = NULL;
+    RESULT_GUARD(s2n_hmac_md_from_alg(alg, &digest));
+    RESULT_ENSURE_REF(digest);
+    const char *digest_name = EVP_MD_get0_name(digest);
+    RESULT_ENSURE_REF(digest_name);
+
+    OSSL_PARAM params[] = {
+        S2N_OSSL_PARAM_INT(OSSL_KDF_PARAM_MODE, mode),
+        S2N_OSSL_PARAM_BLOB(OSSL_KDF_PARAM_KEY, key),
+        S2N_OSSL_PARAM_BLOB(OSSL_KDF_PARAM_INFO, info),
+        S2N_OSSL_PARAM_BLOB(OSSL_KDF_PARAM_SALT, salt),
+        /* Casting away the const is safe because providers are forbidden from
+         * modifying any OSSL_PARAM value other than return_size.
+         * Even the examples in the Openssl documentation cast const strings to
+         * non-const void pointers when setting up OSSL_PARAMs.
+         */
+        S2N_OSSL_PARAM_STR(OSSL_KDF_PARAM_DIGEST, (void *) (uintptr_t) digest_name),
+        OSSL_PARAM_END,
+    };
+
+    /* From the HKDF docs (https://docs.openssl.org/3.1/man7/EVP_KDF-HKDF/):
+     * > When using EVP_KDF_HKDF_MODE_EXTRACT_ONLY the keylen parameter must equal
+     * > the size of the intermediate fixed-length pseudorandom key otherwise an
+     * > error will occur.
+     */
+    if (mode == EVP_KDF_HKDF_MODE_EXTRACT_ONLY) {
+        RESULT_GUARD_OSSL(EVP_KDF_CTX_set_params(hkdf_ctx, params), S2N_ERR_HKDF);
+        size_t key_size = EVP_KDF_CTX_get_kdf_size(hkdf_ctx);
```

Comment:
nit: it might be good to check that `EVP_KDF_CTX_get_kdf_size` didn't error:
> EVP_KDF_CTX_get_kdf_size() returns the output size if the algorithm produces a fixed amount of output and SIZE_MAX otherwise. If an error occurs then 0 is returned.
```suggestion
        size_t key_size = EVP_KDF_CTX_get_kdf_size(hkdf_ctx);
        RESULT_ENSURE(key_size > 0, S2N_ERR_HKDF_OUTPUT_SIZE);
```